munit/shared/src/main/scala-2/munit/internal/MacroCompat.scala

```diff
@@ -1,14 +1,90 @@
 package munit.internal
 
 import munit.Clue
+import munit.Compare
 import munit.Location
 import scala.language.experimental.macros
 import scala.reflect.macros.blackbox.Context
 import scala.reflect.macros.TypecheckException
 import scala.reflect.macros.ParseException
+import scala.collection.mutable
 
 object MacroCompat {
 
+  trait CompareMacro {
+
+    /**
+     * Implements equality according to `==` for any two types (including
+     * unrelated types) as long as the compiler option
+     * "-Xmacro-settings:munit.strictEquality" is not enabled.
+     */
+    implicit def defaultCompareImplicit[A, B]: Compare[A, B] =
+      macro compareImpl[A, B]
+  }
+
+  private val strictEqualityFlag = "munit.strictEquality"
+  private var strictEqualityCache: Option[Boolean] = None
+  private var strictEqualityError: Option[Context => Unit] = None
+  private def isStrictEqualityEnabled(c: Context): Boolean = {
+    strictEqualityCache match {
+      case Some(cachedValue) =>
+        strictEqualityError.foreach(fn => fn(c))
+        cachedValue
+      case _ =>
+        val result = c.settings.contains(strictEqualityFlag)
+        c.settings.foreach { setting =>
+          if (setting.startsWith("munit.") && setting != strictEqualityFlag) {
+            strictEqualityError = Some(context =>
+              context.error(
+                context.enclosingPosition,
+                s"unknown flag '-Xmacro-setting:$setting'. Did you mean -Xmacro-setting:$strictEqualityFlag?"
+              )
+            )
+          }
+        }
+        strictEqualityCache = Some(result)
+        isStrictEqualityEnabled(c)
+    }
+  }
+
+  def compareImpl[A: c.WeakTypeTag, B: c.WeakTypeTag](
+      c: Context
+  ): c.Expr[Compare[A, B]] = {
+    import c.universe._
+    if (!isStrictEqualityEnabled(c)) {
+      reify(Compare.defaultCompare[A, B])
+    } else {
+      val A = weakTypeOf[A]
+      val B = weakTypeOf[B]
+      val solutions = mutable.ListBuffer.empty[String]
+      val lineContent = c.enclosingPosition.lineContent
+      solutions += s"provide an implicit instance of type Equality[$A, $B]"
+
+      if (A =:= B) ()
+      else if (lineContent.contains("assertEquals"))
+        solutions += "use assertEquals[Any, Any](...) if you think it's OK to compare these types at runtime"
+      else if (lineContent.contains("assertNotEquals"))
+        solutions += "use assertNotEquals[Any, Any](...) if you think it's OK to compare these types at runtime"
+
+      solutions += s"""disable strict equality by removing the compiler option "-Xmacro-settings:${strictEqualityFlag}""""
+
+      val allSolutions = solutions.zipWithIndex
+        .map { case (msg, i) => s"  Alternative ${i + 1}: $msg" }
+        .mkString("\n")
+
+      c.error(
```

Comment:
this is probably me not know macros at all, but doesn't this macro always report an error when strict equality is enabled? Where is the code path that reifies the correct `Compare[A, B]` instance?

Reply:
This macro is never called if you have a custom `Compare[A, B]` in scope. I added a comment
```
      // NOTE: report compile error because the user has enabled strict equality
      // and implicit search still picked the default instance in the
      // low-priority `CompareMacro` trait.
```

Reply:
Ahh, makes sense. Thanks!